### PR TITLE
add contest table

### PR DESCRIPTION
**Pull request by drowsy810301**
categorize the typeof contest and let the data of upcoming/running contest depends on db queries

### CODE REVIEW

File: contest/models.py [hunk 1]
```diff
@@ -39,6 +40,22 @@ class Contest(models.Model):
...
+            deltatime = self.start_time  - present
+
+        days, seconds = deltatime.days, deltatime.seconds
+        hours = seconds // 3600
+        minutes = (seconds % 3600) // 60
+        seconds = (seconds % 60)
+        return '{} minute{}, {} hour{}, {} day{}'. \
+            format(minutes, 's' if minutes != 1 else '', \
+                hours, 's' if hours != 1 else '', \
+                days, 's' if days != 1 else '')
```
Comment (nickelcheng) [inline contest/models.py]: You can use pluralize tag.

File: users/models.py [hunk 2]
```diff
@@ -61,6 +61,19 @@ class User(models.Model):
     )
     theme = models.CharField(max_length=8, choices=THEME_CHOICE, default=PAPER)
 
+    def has_admin_auth(self):
+        has_auth = (self.user_level == 'ADMIN')
```
Comment (nickelcheng) [inline users/models.py]: I prefer user_level == ADMIN instead of 'ADMIN'. (variable v.s. hardcode)

File: users/models.py [hunk 3]
```diff
@@ -61,6 +61,19 @@ class User(models.Model):
...
+        return has_auth
+
+    def has_judge_auth(self):
+        has_auth = ((self.user_level == 'ADMIN') or ( self.user_level == 'JUDGE'))
```
Comment (nickelcheng) [inline users/models.py]: Ditto.

File: users/models.py [hunk 4]
```diff
@@ -61,6 +61,19 @@ class User(models.Model):
...
+        return has_auth
+
+    def has_subjudge_auth(self):
+        has_auth = ((self.user_level == 'ADMIN') or ( self.user_level == 'JUDGE') \
```
Comment (nickelcheng) [inline users/models.py]: Ditto.

File: index/views.py [hunk 5]
```diff
@@ -72,3 +85,24 @@ def custom_proc(request):
...
+        if j < present and endlists[i] > present:
+            isRunning.append(endlists[i])
+    return isRunning
+
+def isUpcoming_filter(notFisnished, isRunning):
+    isUpcoming = []
+    for i in notFisnished:
+        if i not in isRunning:
+            isUpcoming.append(i)
+    return isUpcoming
```
Comment (nickelcheng) [inline index/views.py]: I think we don't need these filters?

File: index/views.py [hunk 6]
```diff
@@ -58,9 +69,11 @@ def group_list(request):
     return render(request, 'index/group_list.html',{},
                 context_instance = RequestContext(request, processors = [custom_proc]))
 
-def custom_proc(request):    
-    vol = 0    
-    vol = random.randint(1,10)
+def custom_proc(request):
+    vol = []
+    for i in range(1,11):
```
Comment (nickelcheng) [inline index/views.py]: Please query total number instead of random generating.

File: contest/models.py [hunk 7]
```diff
@@ -39,6 +40,22 @@ class Contest(models.Model):
     open_register = models.BooleanField(default=True)
     creation_time = models.DateTimeField(default=datetime.now, auto_now_add=True)
 
+    def get_time(self):
```
Comment (nickelcheng) [inline contest/models.py]: Maybe use time_diff to avoid confusion?

File: index/static/js/index.js [hunk 8]
```diff
@@ -25,59 +25,6 @@ var displayInfo = false;
...
-            contest_event[Math.floor(Math.random() * contest_num)] + 
-            '</td>' + '<td>' + '</td>' + 
-            '<td>' + Math.floor((Math.random() * 10) + 1) + 
-            ' ' + contest_time[Math.floor(Math.random() * contest_num)] + 
-            '</td>' ;
-        }
-    }
     alert("You have new messages!");
 }
 function info() {
```
Comment (nickelcheng) [inline index/static/js/index.js]: You can refer to https://docs.djangoproject.com/en/dev/ref/templates/builtins/?from=olddocs#now to get current server time.

File: index/views.py [hunk 9]
```diff
@@ -26,15 +26,26 @@
...
+    present = timezone.now()
+    contests = Contest.objects.all()
+    start_times = map(lambda contest: contest.start_time, contests)
+    end_times = map(lambda contest: contest.end_time, contests)
+
+    notFisnished = endcontest_filter(end_times, present)
+    isRunning = isRunning_filter(start_times, end_times, present)
+    isUpcoming = isUpcoming_filter(notFisnished, isRunning)
+
+    cRunnings = Contest.objects.filter(end_time__in=isRunning)
```
Comment (nickelcheng) [inline index/views.py]: Please use same naming rule. Use c_runnings instead.

File: index/views.py [hunk 10]
```diff
@@ -26,15 +26,26 @@
...
+    cUpcomings = Contest.objects.filter(end_time__in=isUpcoming)
```
Comment (nickelcheng) [inline index/views.py]: Ditto.

**Comment by drowsy810301** [comment]
I think I fix all the problem!

File: nthuoj/settings.py [hunk 11]
```diff
@@ -78,7 +78,7 @@
 
 LANGUAGE_CODE = 'en-us'
 
-TIME_ZONE = 'UTC'
+TIME_ZONE = 'Asia/Shanghai'
```
Comment (nickelcheng) [inline nthuoj/settings.py]: Why shanghai?

**Comment by drowsy810301** [comment]
I think I fix all the problem again!

File: index/views.py [hunk 12]
```diff
@@ -58,15 +61,20 @@ def group_list(request):
     return render(request, 'index/group_list.html',{},
                 context_instance = RequestContext(request, processors = [custom_proc]))
 
-def custom_proc(request):    
-    vol = 0    
-    vol = random.randint(1,10)
+def custom_proc(request):
+    volumes = []
+    problems = Problem.objects.all().order_by('id')
+    ids = map(lambda problem: problem.id, problems)
```
Comment (nickelcheng) [inline index/views.py]: You can use https://docs.djangoproject.com/en/1.7/ref/models/querysets/#count instead of counting by yourself.

**Comment by nickelcheng** [comment]
This PR is good for me.
If @bruce3557 and @CHIAHSINHOU don't have other suggestions, I think it can be merged.

**Comment by CHIAHSINHOU** [comment]
建議可以換一下滑鼠移上去後的顏色
跟原本顏色太像了

**Comment by bruce3557** [comment]
+1,going to merge

**Pull request merged**